File: julia/task.jl
using DataFrames
using StatsBase
using MLBase
import XGBoost

include("Lib.jl")
using Lib

srand(42)

fullset = begin
    set = readtable("train.csv")
    set[2:end]
end

applycol!(x -> x - minimum(x) + 1, fullset)

# Seemingly arbitrary ¯\_(ツ)_/¯
const logscale_border = 1e4
applycol!(x -> maximum(x) > logscale_border ? log(x) : x, fullset)

applycol!(uninormalize, fullset)

trainset[end] = map(Int, trainset[end])

(trainset, testset) = train_test_split(fullset, test_size=0.25)
train_n = size(trainset)[1]

# First and, well, the best idea. Oh well...
function est_model(ind)
    nrounds = 1
    XGBoost.xgboost(convert(Matrix, trainset[ind, 1:end-1]), nrounds, label=convert(Array, trainset[ind, end]), max_depth=3, n_estimators=300, learning_rate=0.05)
end

function eval_model(model, ind)
    r = XGBoost.predict(model, convert(Matrix, trainset[ind, 1:end-1]))
    pts = roc(convert(Array, trainset[ind, end]), r, 0:step:1)
    
end

scores = cross_validate(est_model, eval_model, train_n, Kfold(train_n, 5))

File: python/task.py
#!/usr/bin/env python3

import numpy as np
import pandas
from sklearn import cross_validation
from sklearn.cross_validation import train_test_split
from sklearn.metrics import roc_curve, auc
from sklearn import preprocessing
import matplotlib
import xgboost

def pandas_wrap(op, arr):
  return pandas.DataFrame(op(arr), index=arr.index, columns=arr.columns)

np.random.seed(42)
matplotlib.use("QT5Agg")
from matplotlib import pyplot as plt

fullset = pandas.read_csv("train.csv").drop("ID", axis=1)
fullresult = fullset["TARGET"]
fullset.drop("TARGET", axis=1, inplace=True)

# Normalize
fullset = fullset.apply(lambda x: x - x.min() + 1)

# Seemingly arbitrary ¯\_(ツ)_/¯
LOGSCALE_BORDER = 1e4
# Ugh... ternaries in Python
fullset = fullset.apply(lambda x: np.log10(x) if x.max() > LOGSCALE_BORDER else x)

fullset = pandas_wrap(preprocessing.scale, fullset)

trainset, testset, trainresult, testresult = train_test_split(fullset, fullresult, test_size=0.25)

classifier = xgboost.XGBClassifier(max_depth=3)

model = classifier.fit(trainset, trainresult)
probas = model.predict_proba(testset)
fpr, tpr, thresholds = roc_curve(testresult, probas[:, 1])
result_auc = auc(fpr, tpr)
plt.plot(fpr, tpr)
plt.show()

evalset = pandas.read_csv("test.csv")
evalresult = evalset[['ID']]
evalset = evalset.drop('ID', axis=1)

evalresult.insert(1, 'TARGET', model.predict(evalset))
evalresult.to_csv('result.csv', index=False)